### PR TITLE
Migrate Stripe subscriptions to Braintree

### PR DESCRIPTION
**Pull request by cadecairos**
This PR adds an extra step on the charge.succeeded webhook endpoint that will cancel a subscription if one exists, and recreates it on Braintree.

### CODE REVIEW

**Comment by cadecairos** [comment]
proposal: Let's do some integration testing on Staging, since this is behind a feature flag (off by default)

File: donate/payments/tasks.py [hunk 1]
```diff
@@ -245,6 +252,82 @@ def process_charge_succeeded(self, event):
...
+            return
+
+        # Grab the Payment Method token that represents the customers payment information in the Braintree Vault
+        payment_method_token = BT_customer.payment_methods[0].token
+
+        # Determine the Braintree plan, merchant account, and price for the subscription
+        # based on the Stripe subscription
+        BT_plan_id = get_plan_id(subscription.plan.currency)
+        BT_merchant_account = get_merchant_account_id_for_card(subscription.plan.currency)
+        BT_price = zero_decimal_currency_fix(charge.amount, subscription.plan.currency)
```
Comment (cadecairos) [inline donate/payments/tasks.py]: Note: we 100% want to validate that zero-decimal currency subscriptions properly migrate.

File: donate/payments/tasks.py [hunk 2]
```diff
@@ -245,6 +252,82 @@ def process_charge_succeeded(self, event):
...
+        )
+
+        # Ensure we found a customer record, if not, report it to Sentry as an error
+        if len(BT_customers) == 0:
+            logger.error(f'No Customer record exists in Braintree '
+                         f'for the Stripe customer attached to this charge object:'
+                         f'{charge.id}', exc_info=True)
+            return
+
+        # There should be a single Customer/Card combo on Braintree
```
Comment (Pomax) [inline donate/payments/tasks.py]: Should we also add an `if(len(BT_customers) > 1):` here, with a sentry warning?

Comment (cadecairos) [reply donate/payments/tasks.py]: Added in a check and report for this.

File: donate/payments/tasks.py [hunk 3]
```diff
@@ -331,6 +341,71 @@ def process_charge_failed(event):
...
+                      f' Braintree for this charge: {charge.id}'
+            print(message)
+            logger.error(message, exc_info=True)
+            return
+
+        # Determine the Braintree plan, merchant account, and price
+        # for the subscription based on the subscription's attributes
+        BT_plan_id = get_plan_id(charge.currency)
+        BT_merchant_account = get_merchant_account_id_for_card(charge.currency)
+        BT_price = Decimal(zero_decimal_currency_fix(charge.amount, charge.currency))
```
Comment (Pomax) [inline donate/payments/tasks.py]: yes. very smart.

File: donate/payments/tasks.py [hunk 4]
```diff
@@ -331,6 +341,71 @@ def process_charge_failed(event):
...
+        BT_first_billing_date = datetime.utcfromtimestamp(subscription.current_period_end)
+
+        # Create the Braintree subscription and record the result for analysis
+        BT_create_subscription_result = gateway.subscription.create({
+            'payment_method_token': payment_method.token,
```
Comment (Pomax) [inline donate/payments/tasks.py]: Do we need to verify that `payment_method.token` exists, where we're checking if `payment_method` exists? Or does this exist by definition and we don't need to worry about this prop possibly not existing?

Comment (cadecairos) [reply donate/payments/tasks.py]: Token must exist of there is a payment method object. 👍

**Pull request merged**